usr/sbin/p11sak/p11sak.c

```diff
@@ -7729,6 +7749,215 @@ static CK_RV p11sak_import_dilithium_kyber_pem_data(
+        ERR_print_errors_cb(openssl_err_cb, NULL);
+        return NULL;
+    }
+
+    val_len = i2d_ASN1_OBJECT(obj, &val);
+    if (val_len <= 0) {
+        warnx("i2d_ASN1_OBJECT failed for algorithm '%s'", alg_name);
+        ERR_print_errors_cb(openssl_err_cb, NULL);
+        goto done;
+    }
+
+    oid = find_pqc_by_oid(pqc_oids, val, val_len);
+    if (oid == NULL)
+        warnx("Algorithm '%s' is not supported by p11sak", alg_name);
+
+    OPENSSL_free(val);
+
+done:
+    ASN1_OBJECT_free(obj);
+    return oid;
+}
+
+static CK_RV get_octet_string_param_from_pkey(EVP_PKEY *pkey, const char *param,
+                                              CK_BYTE **key, size_t *key_len)
+{
+    if (EVP_PKEY_get_octet_string_param(pkey, param, NULL, 0, key_len) != 1 ||
+        *key_len == OSSL_PARAM_UNMODIFIED) {
+        warnx("EVP_PKEY_get_octet_string_param failed for '%s'\n", param);
+        ERR_print_errors_cb(openssl_err_cb, NULL);
+        return CKR_FUNCTION_FAILED;
+    }
+
+    *key = calloc(1, *key_len);
+    if (*key == NULL) {
+        warnx("Failed to allocate buffer for '%s'\n", param);
+        return CKR_HOST_MEMORY;
+    }
+
+    if (EVP_PKEY_get_octet_string_param(pkey, param,
+                                        *key, *key_len, key_len) != 1) {
+        warnx("EVP_PKEY_get_octet_string_param failed for '%s'\n", param);
+        ERR_print_errors_cb(openssl_err_cb, NULL);
+        return CKR_FUNCTION_FAILED;
+    }
+
+    return CKR_OK;
+}
+#endif
+
+static CK_RV p11sak_import_dilithium_pkey(const struct p11sak_objtype *keytype,
+                                          EVP_PKEY *pkey, bool private,
+                                          CK_ATTRIBUTE **attrs,
+                                          CK_ULONG *num_attrs)
+{
+#if OPENSSL_VERSION_PREREQ(3, 0)
+    const struct pqc_oid *oid;
+    size_t priv_len = 0, pub_len = 0;
+    CK_BYTE *priv_key = NULL, *pub_key = NULL;
+    CK_ULONG ofs;
+    CK_RV rc;
+
+    UNUSED(keytype);
+
+    oid = pqc_oid_from_pkey(pkey, dilithium_oids);
+    if (oid == NULL)
+        return CKR_FUNCTION_FAILED;
+
+    /* Add keyform and attribute */
+    rc = add_attribute(CKA_IBM_DILITHIUM_KEYFORM,
+                       &oid->keyform, sizeof(oid->keyform),
+                       attrs, num_attrs);
+    if (rc != CKR_OK)
+       goto out;
+
+    if (private) {
+        rc = get_octet_string_param_from_pkey(pkey, OSSL_PKEY_PARAM_PRIV_KEY,
+                                              &priv_key, &priv_len);
+        if (rc != CKR_OK)
+            goto out;
+
+        if (priv_len != oid->len_info.dilithium.rho_len +
+                        oid->len_info.dilithium.seed_len +
+                        oid->len_info.dilithium.tr_len +
+                        oid->len_info.dilithium.s1_len +
+                        oid->len_info.dilithium.s2_len +
+                        oid->len_info.dilithium.t0_len) {
+            warnx("Size of private Dilithium key is not valid.");
```

Comment:
Hm, we could give more info like "Size %ld of private Dilithium key is not valid, expected %ld." First insert would be priv_len, second one the sum of the components.

Reply:
Well, if this happens, something is totally wrong. Not sure if knowing the actual and expected sizes would really help.